### PR TITLE
graph-builder: cache release metadata lookups

### PR DESCRIPTION
**Pull request by lucab**
This adds a caching layer for release metadata lookups (both positive
and negative results).

Closes: https://jira.coreos.com/browse/CORS-950

### CODE REVIEW

**Review by steveej — COMMENTED** [review]
Please point me to the negative caching implementation as I can't find it, even though the comments state it's there.

Furthermore you've assumed that tags are immutable, but since they have multiple layers I would assume otherwise. 
Following that the `tag` can't be used as the key in the `HashMap`, and instead the identity (hash) of the `layer_digests` vector should be used, which will change if a tag receives another layer in subsequent runs.
Alternatively we could cache the metadata for each `layer_digest` and lookup there before fetching and investigating the layers.

File: graph-builder/src/registry.rs [hunk 1]
```diff
@@ -114,21 +116,67 @@ pub fn fetch_releases(
...
+        registry_host,
+        repo,
+        tag,
+    );
+    let (tag, release) = thread_runtime
+        .block_on(tagged_release)
+        .context("failed to find first release")?;
+
+    trace!("Caching release metadata for new tag {}", &tag);
+    cache.insert(tag, release.clone());
```
Comment (lucab) [inline graph-builder/src/registry.rs]: If I didn't misread the `find_first_release` logic, a negative result gets here as a `None` release which is then cached for that tag.

Comment (lucab) [reply graph-builder/src/registry.rs]: I will add a docstring on the data-structure to make this more explicit.

Comment (steveej) [reply graph-builder/src/registry.rs]: I see, that makes sense. I missed that `cache.get(..)` wraps it so we're getting `Option<Option<Release>>`.

**Comment by steveej** [comment]
> If tags are immutable, 

Well, they aren't ;-)

> caching them directly will cut on each scrape a number of API calls to quay linear in the number of tags (to fetch every single manifest).

I wouldn't expect too much traffic here as the number of tags per repository will go up slowly as they supposedly correspond to releases.

File: graph-builder/src/registry.rs [hunk 2]
```diff
@@ -163,13 +224,13 @@ fn find_first_release(
     authenticated_client: dkregistry::v2::Client,
     registry_host: String,
     repo: String,
-    tag: String,
-) -> impl Future<Item = Option<Release>, Error = Error> {
-    let tag_for_error = tag.clone();
+    repo_tag: String,
```
Comment (steveej) [inline graph-builder/src/registry.rs:227]: this seems a leftover from your local changes. let's just keep the current names

**Review by steveej — CHANGES_REQUESTED** [review]
Blockers:
* The tests for `fetch_releases` need to be updated
* CI is stuck, I'll look after that

Non-blockers:
* A namechange nitpick in the comments
* We're missing an explicit test for the cache. I'm fine adding that after this PR
* There's one leftover TODO in the code, did you want to get rid of that now or later? I'm fine with keeping it in

File: graph-builder/src/registry.rs (hunk 2 above)
Comment (lucab) [reply graph-builder/src/registry.rs:227]: No, this is to disambiguate between input and output `tag`.

**Comment by lucab** [comment]
@steveeJ I think you reviewed a stale version of this PR, because unless I'm missing something `fetch_releases` tests are already updated (and pass locally).

**Comment by steveej** [comment]
> @steveeJ I think you reviewed a stale version of this PR, because unless I'm missing something `fetch_releases` tests are already updated (and pass locally).

Pardon me! :pray:

/lgtm

**Comment by steveej** [comment]
[test]

**Comment by lucab** [comment]
@steveeJ np. The TODO I left in may require some invasive changes, I'm still researching how to approach it.

**Pull request merged**